From 72581528b867b04c667d6510dd7048ea83f4456a Mon Sep 17 00:00:00 2001
From: "jinshan.xiong" <jinshan.xiong@clusterstor.com>
Date: Thu, 5 Aug 2010 11:09:33 -0600
Subject: [PATCH 28/34] - polish the free of rdma write and reply list, in this
 way, we don't need to export xdr functions for rdma_write_chunk and
 rdma_write_list

---
 net/usunrpc/rfc5666/rfc5666.h     |  2 --
 net/usunrpc/rfc5666/rfc5666_xdr.c |  4 ++--
 net/usunrpc/rfc5666/svc_rdma.c    | 17 ++++++-----------
 3 files changed, 8 insertions(+), 15 deletions(-)

diff --git a/net/usunrpc/rfc5666/rfc5666.h b/net/usunrpc/rfc5666/rfc5666.h
index 5245862..5fa3a81 100644
--- a/net/usunrpc/rfc5666/rfc5666.h
+++ b/net/usunrpc/rfc5666/rfc5666.h
@@ -112,8 +112,6 @@ struct svcrdma_msg {
 
 /* the xdr functions */
 extern bool_t xdr_svcrdma_msg (XDR *, struct svcrdma_msg *);
-extern bool_t xdr_rdma_write_chunk (XDR *, struct rdma_write_chunk *);
-extern bool_t xdr_rdma_write_list (XDR *, struct rdma_write_list *);
 typedef int (*dfunc_t)(const char *, ...);
 extern  void   svcrdma_msg_dump(const char *, struct svcrdma_msg *, dfunc_t);
 
diff --git a/net/usunrpc/rfc5666/rfc5666_xdr.c b/net/usunrpc/rfc5666/rfc5666_xdr.c
index 924167a..77107f0 100644
--- a/net/usunrpc/rfc5666/rfc5666_xdr.c
+++ b/net/usunrpc/rfc5666/rfc5666_xdr.c
@@ -32,7 +32,7 @@ static bool_t xdr_rdma_read_list (XDR *xdrs, struct rdma_read_list *objp)
         return TRUE;
 }
 
-bool_t xdr_rdma_write_chunk (XDR *xdrs, struct rdma_write_chunk *objp)
+static bool_t xdr_rdma_write_chunk (XDR *xdrs, struct rdma_write_chunk *objp)
 {
          if (!xdr_array(xdrs,
                         (char **) &objp->rwc_segs,
@@ -44,7 +44,7 @@ bool_t xdr_rdma_write_chunk (XDR *xdrs, struct rdma_write_chunk *objp)
         return TRUE;
 }
 
-bool_t xdr_rdma_write_list (XDR *xdrs, struct rdma_write_list *objp)
+static bool_t xdr_rdma_write_list (XDR *xdrs, struct rdma_write_list *objp)
 {
         if (!xdr_rdma_write_chunk (xdrs, &objp->rwl_entry))
                 return FALSE;
diff --git a/net/usunrpc/rfc5666/svc_rdma.c b/net/usunrpc/rfc5666/svc_rdma.c
index 773c11d..b0c51c1 100644
--- a/net/usunrpc/rfc5666/svc_rdma.c
+++ b/net/usunrpc/rfc5666/svc_rdma.c
@@ -7,7 +7,6 @@
 #include "svc_rdma_internal.h"
 #include "svc_rdma.h"
 
-#include <err.h>
 #define rdma_errx(rc, fmt, args...)                     \
         fprintf(stderr, "[%s:%d error(%d)] " fmt,       \
                  __func__, __LINE__, rc, ##args)
@@ -842,12 +841,6 @@ static void rdma_reply_put(struct rdma_reply *reply)
 
         xdrmem_create(&xdrs, NULL, 0, XDR_FREE);
         (void)xdr_svcrdma_msg(&xdrs, &reply->rp_msg);
-        (void)xdr_reference(&xdrs, (caddr_t *)&reply->rp_orig_writes,
-                            sizeof(*reply->rp_orig_writes), 
-                            (xdrproc_t)xdr_rdma_write_list);
-        (void)xdr_reference(&xdrs, (caddr_t*)&reply->rp_reply,
-                            sizeof(*reply->rp_reply), 
-                            (xdrproc_t)xdr_rdma_write_chunk);
 
         if (reply->rp_rpc == &reply->rp_rpc_chunks) {
                 rdma_bufdesc_free(xprt, reply->rp_rpc);
@@ -1072,9 +1065,7 @@ static int rdma_recv_completion(struct svcrdma_xprt *xprt,
         /* everything looks ok, migrate chunks from request */
         reply->rp_orig_writes = reply->rp_writes = *req->rq_write_list;
         reply->rp_writes_index = 0;
-        *req->rq_write_list = NULL;
         reply->rp_reply = *req->rq_reply_chunk;
-        *req->rq_reply_chunk = NULL;
 
         /* XXX: we have to use rx_curreq to position current request in the
           coming callbacks. */
@@ -1384,6 +1375,7 @@ static int rdma_reply_encode(struct rdma_reply *reply)
         int bufsize;
         XDR xdrs;
 
+        C2_ASSERT(reply->rp_request != NULL);
         memset(msg, 0, sizeof(*msg));
         msg->rdma_xid    = reply->rp_xid;
         msg->rdma_vers   = RDMA_VERS;
@@ -1426,8 +1418,11 @@ static int rdma_reply_encode(struct rdma_reply *reply)
                         rdma_nomsg->rdma_reply = reply->rp_reply;
                         reply->rp_rpc_buflen = 0;
                 }
-                reply->rp_orig_writes = NULL;
-                reply->rp_reply       = NULL;
+                /* set the pointers to be NULL so that it won't be freed
+                 by XDR_FREE in rdma_request_repost(). Instead, they will
+                 be freed in rdma_reply_put. */
+                *reply->rp_request->rq_write_list  = NULL;
+                *reply->rp_request->rq_reply_chunk = NULL;
         }
 
         bufsize = reply->rp_rdma.rbd_bufsz;
-- 
1.8.3.2

